### PR TITLE
Initial plumbing for metrics.

### PR DESCRIPTION
**Pull request by boks1971**
Reviewers: Looking for comments on the flow/structure. There is a bunch more to do. But, wanted to ensure that the plumbing and the approach looks okay before proceeding further.

This implements
- metrics received from participant.
- server time stamp added based on one-way-delay estimation.
- callback to room.
- room distributes it to all  participants including the sending participant
- other participants forward to client.
- counting metrics bytes in data channel stats

TODO:
  - recording/processing/batching - would be for analytics/off line processing - coming later
  - should recording/processing/batching happen on publisher side or subscriber side?
  - should metrics be echoed back to publisher? Selectively, i. e. subscriber side stats like freeze duration etc. would be sent back to the sender
  - ~grants to publish/subscribe metrics.~ - no grants needed for publish, added grants for subscribe side in https://github.com/livekit/protocol/pull/795)

### CODE REVIEW

File: test/client/client.go [hunk 1]
```diff
@@ -700,7 +700,6 @@ func (c *RTCClient) PublishData(data []byte, kind livekit.DataPacket_Kind) error
 	}
 
 	dpData, err := proto.Marshal(&livekit.DataPacket{
-		Kind: kind,
```
Comment (boks1971) [inline test/client/client.go:703]: Removing deprecated field.

File: pkg/rtc/participant.go [hunk 2]
```diff
@@ -1595,26 +1626,40 @@ func (p *ParticipantImpl) onDataMessage(kind livekit.DataPacket_Kind, data []byt
...
 		}
 	}
+	if shouldForwardMetrics {
+		if onMetrics := p.getOnMetrics(); onMetrics != nil {
+			onMetrics(p, dp)
+		}
+	}
 
-	p.setIsPublisher(true)
+	if isPublisher {
```
Comment (boks1971) [inline pkg/rtc/participant.go:1660]: Not treating `metrics` sender as `publisher`. Should that not be?

File: pkg/rtc/participant.go [hunk 3]
```diff
@@ -2640,16 +2685,14 @@ func (p *ParticipantImpl) SendDataPacket(kind livekit.DataPacket_Kind, encoded [
 		return ErrDataChannelUnavailable
 	}
 
-	err := p.TransportManager.SendDataPacket(kind, encoded)
-	if err != nil {
-		if (errors.Is(err, sctp.ErrStreamClosed) || errors.Is(err, io.ErrClosedPipe)) && p.params.ReconnectOnDataChannelError {
```
Comment (boks1971) [inline pkg/rtc/participant.go:2645]: moving error check/logging to `TransportManager` to keep it in one place.

**Comment by boks1971** [comment]
@reviewers, marking this ready for review. This will forward metrics `as-is` to participants with `CanSubscribeMetrics` permissions. Please let me know what you think about `METRICS-TODO-QUESTIONS` part, especially the part about destination identities. Do we need that?

I think we can ignore the batching part for now. We can discuss that when figuring out sending to analytics for offline processing.

Another thing is time stamps. Want to get this pipeline flowing first before doing more work. Would really like to understand what real time clients are going to do with this and figure out how to deal with time stamps.

File: pkg/rtc/mediatrack.go [hunk 4]
```diff
@@ -56,6 +56,7 @@ type MediaTrack struct {
 }
 
 type MediaTrackParams struct {
+	BaseTime              time.Time
```
Comment (paulwe) [inline pkg/rtc/mediatrack.go:59]: did we end up making a global for this?

File: pkg/rtc/room.go [hunk 5]
```diff
@@ -1792,14 +1791,21 @@ func BroadcastDataPacketForRoom(r types.Room, source types.LocalParticipant, kin
...
-		err := op.SendDataPacket(kind, dpData)
-		if err != nil && !errors.Is(err, io.ErrClosedPipe) && !errors.Is(err, sctp.ErrStreamClosed) &&
-			!errors.Is(err, ErrTransportFailure) && !errors.Is(err, ErrDataChannelBufferFull) {
-			op.GetLogger().Infow("send data packet error", "error", err)
-		}
+		op.SendDataPacket(kind, dpData)
 	})
 }
 
+func BroadcastMetricsForRoom(r types.Room, source types.LocalParticipant, dp *livekit.DataPacket, logger logger.Logger) {
```
Comment (paulwe) [inline pkg/rtc/room.go:1798]: is this meant to be called from outside rtc?

File: pkg/rtc/transportmanager.go [hunk 6]
```diff
@@ -241,7 +245,19 @@ func (t *TransportManager) RemoveSubscribedTrack(subTrack types.SubscribedTrack)
 
 func (t *TransportManager) SendDataPacket(kind livekit.DataPacket_Kind, encoded []byte) error {
 	// downstream data is sent via primary peer connection
-	return t.getTransport(true).SendDataPacket(kind, encoded)
+	err := t.getTransport(true).SendDataPacket(kind, encoded)
+	if err != nil {
+		if !errors.Is(err, io.ErrClosedPipe) && !errors.Is(err, sctp.ErrStreamClosed) && !errors.Is(err, ErrTransportFailure) && !errors.Is(err, ErrDataChannelBufferFull) {
```
Comment (paulwe) [inline pkg/rtc/transportmanager.go:250]: `!cloudutils.ErrorIsOneOf(err, etc...)`?

**Review by paulwe — APPROVED** [review]

File: pkg/rtc/mediatrack.go (hunk 4 above)
Comment (boks1971) [reply pkg/rtc/mediatrack.go:59]: Not yet. Haven't gone back to look at it. Been lagging :-(

File: pkg/rtc/room.go (hunk 5 above)
Comment (boks1971) [reply pkg/rtc/room.go:1798]: `cloudrtc` will use it too.

File: pkg/rtc/transportmanager.go (hunk 6 above)
Comment (boks1971) [reply pkg/rtc/transportmanager.go:250]: Need to move that to OSS somewhere.

Comment (paulwe) [reply pkg/rtc/transportmanager.go:250]: ...forgot where we were

**Pull request merged**